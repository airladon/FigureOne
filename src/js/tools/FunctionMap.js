// @flow
// import type { Transform } from '../tools/g2';
// import { Point, getTransform, Transform } from '../tools/g2';
// import { round } from '../tools/math';
// import type { FigureElement } from './Element';
// import TimeKeeper from './webgl/TimeKeeper';

// Singleton class that contains projects global functions

class GeneralFunctionMap {
  map: { [id: string]: {
    fn: Function,
  }}

  constructor() {
    this.map = {};
  }

  add(id: string, fn: Function) {
    if (this.map == null) {
      this.map = {};
    }
    this.map[id] = {
      fn,
    };
  }

  exec(idOrFn: string | Function | null, ...args: any) {
    if (idOrFn == null) {
      return null;
    }
    if (typeof idOrFn === 'string') {
      // return this.fnMap.exec(fn, ...args);
      if (this.map[idOrFn] != null) {
        if (args.length === 0) {
          return this.map[idOrFn].fn();
        }
        return this.map[idOrFn].fn(...args);
      }
    }
    if (typeof idOrFn === 'function') {
      return idOrFn(...args);
    }
    throw new Error(`${idOrFn} needs to be a function or a string`);
  }

  execOnMaps(idOrFn: string | Function | null, mapsIn: Array<Object>, ...args: any) {
    if (idOrFn == null) {
      return null;
    }
    if (typeof idOrFn === 'string') {
      const maps = [this.map, ...mapsIn];
      for (let i = 0; i < maps.length; i += 1) {
        const map = maps[i];
        if (map[idOrFn] != null) {
          if (args.length === 0) {
            return map[idOrFn].fn();
          }
          return map[idOrFn].fn(...args);
        }
      }
      throw new Error(`FunctionMap Error: '${idOrFn}' does not exist in map`);
    }
    if (typeof idOrFn === 'function') {
      return idOrFn(...args);
    }
    throw new Error(`FunctionMap Error: '${idOrFn}' needs to be a function or a string`);
    // return null;
  }
}


class GlobalFunctionMap extends GeneralFunctionMap {
  static instance: Object;
  constructor() {
    // If the instance alread exists, then don't create a new instance.
    // If it doesn't, then setup some default values.
    if (!GlobalFunctionMap.instance) {
      super();
      GlobalFunctionMap.instance = this;
      this.map = {};
    }
    return GlobalFunctionMap.instance;
  }
}

/**
 * Function Map
 *
 * In FigureOne {@link Recorder}, state is saved in stringified javascript
 * objects. When the state includes a function (like a trigger method in an
 * animation for example) then that function is referenced in the state object
 * as a unique string id.
 *
 * When a geometry is loaded, functions that will be captured in state objects
 * need to be added to a function map. Both {@link Figure} and
 * {@link FigureElement} have attached function maps as the property `fnMap`.
 * Therefore, the method is added to either the figure or element function map,
 * with an associated unique string id, and that string id is used when the
 * function is used. For example as defined callbacks, triggerAnimationStep and
 * customAnimationStep callbacks, or recorder slide definitions (`entryState`,
 * `steadyState`, `leaveState`, `exec`, `execDelta`).
 *
 * The funciton map has:
 * - A map of functions
 * - A link to a global map of functions (a singleton)
 * - The ability to execute a function within the map
 *
 * @property {FunctionMap} global global function map
 * @property {Object} map local function map where keys are unique identifiers
 * and values are the associated functions
 *
 * @example
 * // Add a console function to a FunctionMap and execute it with a parameter
 * figure.fnMap.add('toConsole', s => console.log(s));
 * figure.fnMap.exec('toConsole', 'hello');
 */
class FunctionMap extends GeneralFunctionMap {
  map: { [id: string]: {
    fn: Function,
  }}

  global: GlobalFunctionMap;

  constructor() {
    super();
    this.global = new GlobalFunctionMap();
  }

  /**
   * Execute function with arguments.
   *
   * @param {string | Function | null} idOrFn If `string`, then a function
   * in the local map or global map will be executed (local map takes
   * precedence). If `Function`, then the
   * function will be exectuted. If `null`, then nothing will happen.
   * @param {any} ...args arguments to pass to function.
   */
  exec(idOrFn: string | Function | null, ...args: any) {
    return this.execOnMaps(idOrFn, [this.global.map], ...args);
  }

  execOnMaps(idOrFn: string | Function | null, mapsIn: Array<Object>, ...args: any) {
    return super.execOnMaps(idOrFn, [...mapsIn, this.global.map], ...args);
  }

  /**
   * Add a function to the map.
   * @param {string} id unique identifier
   * @param {Function} fn function to add
   */
  add(id: string, fn: Function) {
    return super.add(id, fn);
  }
}

export { FunctionMap, GlobalFunctionMap };
