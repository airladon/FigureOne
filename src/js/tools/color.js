// @flow

import colorNames from './colorNames';  // official css color names
// eslint-disable-next-line import/no-cycle
import getCSSColors from './getCssColors';
import { round } from './math';
// Function that converts any rgb or rgba string to an array of rgba numbers
// between 0 and 1
function RGBToArray(color: string): Array<number> {
  // Reduce the rgb(a) string to just numbers
  let colString: string = color;
  colString = colString.replace(/.*\(/i, '');
  colString = colString.replace(/\)/i, '');
  const strArray = colString.split(',');
  // Go through each rgb(a) value and normalize to 1.0
  const value: Array<number> = strArray.map((x, index) => {
    if (index < 3) {
      return parseInt(x, 10) / 255.0;
    }
    return parseFloat(x);
  });

  // If an alpha value isn't included, then include it with default value 1.0
  if (value.length === 3) {
    value.push(1);
  }
  return value;
}


// Function that converts any hex color string to an array of rgba numbers
// between 0 and 1 (where alpha is always 1)
function HexToArray(color: string): Array<number> {
  let colHex: string = color.slice(1);
  if (colHex.length < 6) {
    colHex =
      `${colHex[0]}${colHex[0]}${colHex[1]}${colHex[1]}${colHex[2]}${colHex[2]}`;
  }
  const col: Array<number> = [
    parseInt(colHex.slice(0, 2), 16) / 255.0,
    parseInt(colHex.slice(2, 4), 16) / 255.0,
    parseInt(colHex.slice(4, 6), 16) / 255.0,
    1,
  ];
  return col;
}

function cssColorToArray(cssColorString: string): Array<number> | null {
  const oNames = colorNames();               // Official css color names
  let color = cssColorString.slice(0);

  // If the color is an official name, then replace it with the hex rgb
  // equivalent
  if (color in oNames) {
    color = oNames[color];
  }

  // colorValue is the rgba array of colors between 0 and 1
  let colorValue: Array<number> = [];

  // If color string starts with 'rgb' (and therefore also 'rgba')
  if (color.toLowerCase().startsWith('rgb')) {
    colorValue = RGBToArray(color);

  // If color string starts with '#' it is hex
  } else if (color.startsWith('#')) {
    colorValue = HexToArray(color);
  }

  // If the color value array is defined, then add it to the final
  // dictionary
  if (colorValue.length > 0) {
    return colorValue;
  }
  return null;
}

function colorArrayToRGBA(color: TypeColor) {
  return `rgba(${
    Math.floor(color[0] * 255)},${
    Math.floor(color[1] * 255)},${
    Math.floor(color[2] * 255)},${
    color[3]})`;
}

function colorArrayToRGB(color: TypeColor) {
  return `rgb(${
    Math.floor(color[0] * 255)},${
    Math.floor(color[1] * 255)},${
    Math.floor(color[2] * 255)})`;
}

function areColorsSame(color1: Array<number>, color2: Array<number>, precision: number = 5) {
  if (color1.length !== color2.length) {
    return false;
  }
  for (let i = 0; i < color1.length; i += 1) {
    if (round(color1[i], precision) !== round(color2[i], precision)) {
      return false;
    }
  }
  return true;
}

function areColorsWithinDelta(color1: Array<number>, color2: Array<number>, delta: number = 0.001) {
  if (color1.length !== color2.length) {
    return false;
  }
  for (let i = 0; i < color1.length; i += 1) {
    const dC = Math.abs(color1[i] - color2[i]);
    if (dC > delta) {
      return false;
    }
  }
  return true;
}

export {
  RGBToArray, HexToArray, cssColorToArray, colorArrayToRGB,
  colorArrayToRGBA, getCSSColors, areColorsSame, areColorsWithinDelta,
};
