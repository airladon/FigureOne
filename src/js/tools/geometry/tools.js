// @flow
import { Point } from './Point';

function pointsToNumbers2(points: Array<Point>) {
  const list = Array(points.length * 2);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 2] = points[i].x;
    list[i * 2 + 1] = points[i].y;
  }
  return list;
}

function pointsToNumbers(points: Array<Point>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return pointsToNumbers2(points);
  }
  const list = Array(points.length * 3);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 3] = points[i].x;
    list[i * 3 + 1] = points[i].y;
    list[i * 3 + 2] = points[i].z;
  }
  return list;
}

function numbersToPoints2(numbers: Array<number>) {
  const points = Array(numbers.length / 2);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(numbers[i * 2], numbers[i * 2 + 1]);
  }
  return points;
}

function numbersToPoints(numbers: Array<number>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return numbersToPoints2(numbers);
  }
  const points = Array(numbers.length / 3);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(
      numbers[i * 3],
      numbers[i * 3 + 1],
      numbers[i * 3 + 2],
    );
  }
  return points;
}

function toNumbers(
  pointsOrNumbers: Array<Point> | Array<Array<number>> | Array<number>,
  dimension: 2 | 3 = 3,
) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] === 'number') {
    return pointsOrNumbers;
  }
  if (Array.isArray(pointsOrNumbers[0])) {
    const out = [];
    for (let i = 0; i < pointsOrNumbers.length; i += 1) {
      out.push(...pointsOrNumbers);
    }
    return out;
  }
  return pointsToNumbers(pointsOrNumbers, dimension);
}

function toPoints(pointsOrNumbers: Array<Point> | Array<number>, dimension: 2 | 3 = 3) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] !== 'number') {
    return pointsOrNumbers;
  }
  return numbersToPoints(pointsOrNumbers, dimension);
}

export {
  pointsToNumbers2,
  pointsToNumbers,
  numbersToPoints,
  numbersToPoints2,
  toNumbers,
  toPoints,
};
