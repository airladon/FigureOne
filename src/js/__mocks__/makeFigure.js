// import {
//   FigureElementPrimitive,
//   FigureElementCollection,
//   // AnimationPhase,
// } from '../Element';
import Figure from '../figure/Figure';
import {
  Rect, getPoint, Point, getRect,
} from '../tools/g2';
import webgl from './WebGLInstanceMock';
import DrawContext2D from './DrawContext2DMock';
import * as tools from '../tools/tools';
import { round } from '../tools/math';

tools.isTouchDevice = jest.fn();

jest.mock('../figure/Gesture');
jest.mock('../figure/webgl/webgl');
jest.mock('../figure/DrawContext2D');
jest.mock('../figure/DrawingObjects/HTMLObject/HTMLObject');

const generateRandomStringMock = jest.spyOn(tools, 'generateRandomString');
generateRandomStringMock.mockImplementation(() => '000000');

export default function makeFigure(
  inputCanvasIn = new Rect(100, -300, 1000, 500),
  scene = {},
) {
  document.body.innerHTML =
    '<div id="c">'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__low">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__low">'
    + '  </canvas>'
    + '  <div class="figureone__html">'
    + '  </div>'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__high">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__high">'
    + '  </canvas>'
    + '</div>';
  // canvas = document.getElementById('c');
  const inputCanvas = getRect(inputCanvasIn);
  // const inputLimits = getRect(inputLimitsIn);
  const definition = {
    width: inputCanvas.width,
    height: inputCanvas.height,
    limits: new Rect(-1, -1, 2, 2),
  };

  const canvasMock = {
    width: definition.width,
    height: definition.height,
    // offsetLeft: 100,
    left: inputCanvas.left,
    // offsetTop: 200,
    top: inputCanvas.top,
    // width: definition.width,
    // height: definition.height,
    offsetWidth: definition.width,
    offsetHeight: definition.height,
    scrollLeft: 0,
    scrollTop: 0,
    // eslint-disable-next-line arrow-body-style
    getBoundingClientRect: () => {
      return {
        left: inputCanvas.left,
        top: inputCanvas.top,
        width: definition.width,
        height: definition.height,
      };
    },
  };
  const htmlCanvasMock = {
    style: {
      fontsize: 1,
    },
    offsetWidth: 100,
    appendChild: () => {},
  };
  const { limits } = definition;
  const figure = new Figure({
    htmlId: 'c', limits, color: [1, 0, 0, 1], scene,
  });
  figure.webglLow = webgl;
  figure.webglHigh = webgl;
  figure.webgl = webgl;
  figure.canvasLow = canvasMock;
  figure.webglLow.gl.canvas = canvasMock;
  figure.canvasHigh = canvasMock;
  figure.htmlCanvas = htmlCanvasMock;
  figure.isTouchDevice = false;
  figure.draw2DLow = new DrawContext2D(definition.width, definition.height);
  figure.draw2DHigh = new DrawContext2D(definition.width, definition.height);
  figure.draw2D = figure.draw2DLow;
  figure.shapesLow = figure.getShapes(false);
  figure.shapesHigh = figure.getShapes(true);
  figure.shapes = figure.shapesLow;
  figure.primitives = figure.shapes;
  // figure.equationLow = figure.getEquations(false);
  // figure.equationHigh = figure.getEquations(true);
  figure.equation = figure.equationLow;
  figure.collectionsLow = figure.getObjects(false);
  figure.collectionsHigh = figure.getObjects(true);
  figure.collections = figure.collectionsLow;
  figure.createFigureElements();
  figure.defaultLineWidth = 0.01;
  figure.primitives.defaultLineWidth = 0.01;
  // needed as the first element needs to be set with the space Transforms
  figure.initElements();
  figure.mock = {
    initialTime: 0,
    duration: 0,
    previousTouchPoint: new Point(0, 0),
    timersBeforeDraw: true,
    timeStep: (deltaTimeInSecondsIn, frameTimeIn = null, draw = true) => {
      const { duration, initialTime } = figure.mock;
      const deltaTimeInSeconds = round(deltaTimeInSecondsIn, 8);
      let frameTime = deltaTimeInSeconds;
      // let deltaTime = 0;
      if (frameTimeIn != null && frameTimeIn < deltaTimeInSeconds) {
        frameTime = frameTimeIn;
        // deltaTime = frameTime;
      }
      let deltaTime = frameTime;

      // let delta = Math.min(deltaTimeInSeconds, frameTime);
      let lastTime = 0;
      deltaTime = round(deltaTime, 8);
      while (deltaTime <= round(deltaTimeInSeconds, 8)) {
        const newNow = round((duration + deltaTime + initialTime) * 1000, 8);
        global.performance.now = () => newNow;
        if (figure.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        if (draw) {
          figure.draw(round(newNow / 1000, 8));
        }
        if (!figure.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        lastTime = deltaTime;
        if (deltaTime === deltaTimeInSeconds || frameTime === 0) {
          deltaTime += 1;
        } else if (deltaTime + frameTime > deltaTimeInSeconds) {
          deltaTime = deltaTimeInSeconds;
        } else {
          deltaTime += frameTime;
        }
        deltaTime = round(deltaTime, 8);
      }
      figure.mock.duration += deltaTimeInSeconds;
    },
    touchDown: (figurePosition) => {
      const p = getPoint(figurePosition);
      figure.touchDownHandler(p);
      figure.mock.previousTouchPoint = p;
    },
    touchUp: () => {
      figure.touchUpHandler();
    },
    touchMove: (figurePosition) => {
      const p = getPoint(figurePosition);
      figure.touchMoveHandler(figure.mock.previousTouchPoint, p);
      figure.mock.previousTouchPoint = p;
    },
  };
  figure.timeKeeper.reset();
  return figure;
}

