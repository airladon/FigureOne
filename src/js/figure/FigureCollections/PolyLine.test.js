// import {
//   FigureElementPrimitive,
//   FigureElementCollection,
//   // AnimationPhase,
// } from '../Element';
// import Figure from '../Figure';
import {
  Point, rectToPolar,
} from '../../tools/g2';
import {
  round,
} from '../../tools/math';
import * as tools from '../../tools/tools';
// import webgl from '../../__mocks__/WebGLInstanceMock';
// import DrawContext2D from '../../__mocks__/DrawContext2DMock';
import makeFigure from '../../__mocks__/makeFigure';

// tools.isTouchDevice = jest.fn();
jest.useFakeTimers();

// jest.mock('../Gesture');
// jest.mock('../webgl/webgl');
// jest.mock('../DrawContext2D');
// jest.mock('../../tools/tools');

describe('Figure Objects PolyLine', () => {
  let figure;
  let ways;
  let points;

  beforeEach(() => {
    figure = makeFigure();
    points = [
      new Point(0, 0),
      new Point(1, 0),
      new Point(1, 1),
    ];
    ways = {
      numPointsClose: () => figure.collections.polyline({
        points,
        close: true,
      }),
      numPointsOpen: () => figure.collections.polyline({
        points,
        close: false,
      }),
      SideLabelsClose: () => figure.collections.polyline({
        points,
        close: true,
        side: {
          label: {
            text: 'a',
          },
        },
      }),
      SideLabelsOpen: () => figure.collections.polyline({
        points,
        close: false,
        side: {
          label: {
            text: 'a',
          },
        },
      }),
      SideLabelsFullDefine: () => figure.collections.polyline({
        points,
        close: true,
        side: [
          { label: { text: 'a' } },
          { label: { text: 'b' } },
          { label: { text: null } },
        ],
      }),
      SideLabelsUnderDefine: () => figure.collections.polyline({
        points,
        close: true,
        side: [
          { label: { text: 'a' } },
          { label: { text: 'b' } },
        ],
      }),
      SideLabelsOverDefine: () => figure.collections.polyline({
        points,
        close: true,
        side: [
          { label: { text: 'a' } },
          { label: { text: 'b' } },
          { label: { text: 'c' } },
          { label: { text: 'd' } },
        ],
      }),
      SideLabelsSingleDefine: () => figure.collections.polyline({
        points,
        close: true,
        side: { label: { text: 'a' } },
      }),
      SideLabelsActualLength: () => figure.collections.polyline({
        points,
        close: true,
        side: { label: { text: null } },
      }),
      AngleLabelsClose: () => figure.collections.polyline({
        points,
        close: true,
        angle: {
          label: {
            text: 'a',
          },
        },
      }),
      AngleLabelsOpen: () => figure.collections.polyline({
        points,
        close: false,
        angle: {
          label: {
            text: 'a',
          },
        },
      }),
      AngleLabelsFullDefine: () => figure.collections.polyline({
        points,
        close: true,
        angle: [
          { label: { text: 'a' } },
          { label: { text: 'b' } },
          { label: { text: null } },
        ],
      }),
      PadSimple: () => figure.collections.polyline({
        points,
        close: true,
        pad: {
          radius: 0.2,
        },
      }),
      PadBoundary: () => figure.collections.polyline({
        points,
        close: true,
        pad: {
          radius: 0.2,
          boundary: {
            translation: {
              left: -3, bottom: -2, right: 3, top: 2,
            },
          },
          isMovable: true,
        },
      }),
      Misc: () => figure.collections.polyline({
        points,
        close: true,
        color: [1, 0, 0, 1],
        side: {
          color: [1, 0, 0, 1],
          offset: 0.25,
          showLine: true,
          arrows: true,
          width: 0.015,
          label: {
            text: null,
            location: 'outside',
            scale: 0.5,
            color: [0, 1, 0, 1],
          },
          mods: {
            isMovable: true,
            isTouchable: true,
          },
        },
        angle: {
          label: {
            text: null,
            radius: 0.25,
            textScale: 0.5,
            color: [1, 0, 1, 1],
          },
          curve: {
            radius: 0.3,
            sides: 50,
          },
          mods: {
            isMovable: true,
            isTouchable: true,
          },
        },
        pad: {
          color: [1, 0.5, 0.5, 1],
          radius: 0.2,
          isMovable: true,
          touchRadius: 0.4,
        },
      }),
    };
  });
  test('Number of points in line close', () => {
    const poly = ways.numPointsClose();
    // expect(poly._line.drawingObject.points).toHaveLength(points.length * 12);
    expect(tools.cleanUIDs(poly._line)).toMatchSnapshot();
    expect(poly._line.drawingObject.points).toHaveLength(points.length * 12);
    expect(Object.keys(poly.elements)).toHaveLength(1);
  });
  test('Number of points in line open', () => {
    const poly = ways.numPointsOpen();
    expect(poly._line.drawingObject.points).toHaveLength((points.length - 1) * 12);
    expect(Object.keys(poly.elements)).toHaveLength(1);
  });
  test('Misc options', () => {
    const poly = ways.Misc();
    expect(tools.cleanUIDs(poly)).toMatchSnapshot();
  });
  describe('Pads', () => {
    test('Simple', () => {
      const poly = ways.PadSimple();
      expect(poly).toHaveProperty('_pad0');
    });
    test('Boundary', () => {
      const poly = ways.PadBoundary();
      expect(poly).toHaveProperty('_pad0');
      const { boundary } = poly._pad0.getMoveBounds().boundary[0];
      expect(round(boundary.left, 3)).toBe(-2.8);
      expect(round(boundary.right, 3)).toBe(2.8);
      expect(round(boundary.bottom, 3)).toBe(-1.8);
      expect(round(boundary.top, 3)).toBe(1.8);
    });
  });
  describe('Side Labels', () => {
    test('Close', () => {
      const poly = ways.SideLabelsClose();
      expect(Object.keys(poly.elements)).toHaveLength(4);
      expect(Object.keys(poly.elements))
        .toEqual(['line', 'side01', 'side12', 'side20']);
      expect(poly.elements.side01.line.p1).toEqual(points[0]);
      expect(poly.elements.side01.line.p2).toEqual(points[1]);
      expect(poly.elements.side12.line.p1).toEqual(points[1]);
      expect(poly.elements.side12.line.p2).toEqual(points[2]);
      expect(poly.elements.side20.line.p1).toEqual(points[2]);
      expect(poly.elements.side20.line.p2.round()).toEqual(points[0]);
    });
    test('Open', () => {
      const poly = ways.SideLabelsOpen();
      expect(Object.keys(poly.elements)).toHaveLength(3);
      expect(Object.keys(poly.elements))
        .toEqual(['line', 'side01', 'side12']);
      expect(poly.elements.side01.line.p1).toEqual(points[0]);
      expect(poly.elements.side01.line.p2).toEqual(points[1]);
      expect(poly.elements.side12.line.p1).toEqual(points[1]);
      expect(poly.elements.side12.line.p2).toEqual(points[2]);
    });
    test('Full Define', () => {
      const poly = ways.SideLabelsFullDefine();
      const base01 = poly.elements.side01.label.eqn._base;
      const base12 = poly.elements.side12.label.eqn._base;
      const base20 = poly.elements.side20.label.eqn._base;
      expect(base01.drawingObject.text[0].text).toBe('a');
      expect(base12.drawingObject.text[0].text).toBe('b');
      expect(base20.drawingObject.text[0].text).toBe('1.4');
    });
    test('Under Define', () => {
      const poly = ways.SideLabelsUnderDefine();
      const base01 = poly.elements.side01.label.eqn._base;
      const base12 = poly.elements.side12.label.eqn._base;
      const base20 = poly.elements.side20.label.eqn._base;
      expect(base01.drawingObject.text[0].text).toBe('a');
      expect(base12.drawingObject.text[0].text).toBe('b');
      expect(base20.drawingObject.text[0].text).toBe('a');
    });
    test('Over Define', () => {
      const poly = ways.SideLabelsOverDefine();
      const base01 = poly.elements.side01.label.eqn._base;
      const base12 = poly.elements.side12.label.eqn._base;
      const base20 = poly.elements.side20.label.eqn._base;
      expect(base01.drawingObject.text[0].text).toBe('a');
      expect(base12.drawingObject.text[0].text).toBe('b');
      expect(base20.drawingObject.text[0].text).toBe('c');
    });
    test('Single Define', () => {
      const poly = ways.SideLabelsSingleDefine();
      const base01 = poly.elements.side01.label.eqn._base;
      const base12 = poly.elements.side12.label.eqn._base;
      const base20 = poly.elements.side20.label.eqn._base;
      expect(base01.drawingObject.text[0].text).toBe('a');
      expect(base12.drawingObject.text[0].text).toBe('a');
      expect(base20.drawingObject.text[0].text).toBe('a');
    });
    test('Actual Length', () => {
      const poly = ways.SideLabelsActualLength();
      const base01 = poly.elements.side01.label.eqn._base;
      const base12 = poly.elements.side12.label.eqn._base;
      const base20 = poly.elements.side20.label.eqn._base;
      expect(base01.drawingObject.text[0].text).toBe('1.0');
      expect(base12.drawingObject.text[0].text).toBe('1.0');
      expect(base20.drawingObject.text[0].text).toBe('1.4');
    });
  });
  describe('Angle Labels', () => {
    test('Close', () => {
      const poly = ways.AngleLabelsClose();
      expect(Object.keys(poly.elements)).toHaveLength(4);
      expect(Object.keys(poly.elements))
        .toEqual(['angle0', 'angle1', 'angle2', 'line']);
      // expect(poly.elements.angle0.p1).toEqual(points[1]);
      // expect(poly.elements.angle0.p2).toEqual(points[0]);
      // expect(poly.elements.angle0.p3).toEqual(points[2]);
      // expect(poly.elements.side01.p2).toEqual(points[1]);
      // expect(poly.elements.side12.p1).toEqual(points[1]);
      // expect(poly.elements.side12.p2).toEqual(points[2]);
      // expect(poly.elements.side20.p1).toEqual(points[2]);
      // expect(poly.elements.side20.p2.round()).toEqual(points[0]);
    });
    test('Open', () => {
      const poly = ways.AngleLabelsOpen();
      expect(Object.keys(poly.elements)).toHaveLength(2);
      expect(Object.keys(poly.elements))
        .toEqual(['angle0', 'line']);
      // expect(poly.elements.side01.p1).toEqual(points[0]);
      // expect(poly.elements.side01.p2).toEqual(points[1]);
      // expect(poly.elements.side12.p1).toEqual(points[1]);
      // expect(poly.elements.side12.p2).toEqual(points[2]);
    });
    test('Full Define', () => {
      const poly = ways.AngleLabelsFullDefine();
      const base01 = poly.elements.angle0.label.eqn._base;
      const base12 = poly.elements.angle1.label.eqn._base;
      const base20 = poly.elements.angle2.label.eqn._base;
      expect(base01.drawingObject.text[0].text).toBe('a');
      expect(base12.drawingObject.text[0].text).toBe('b');
      expect(base20.drawingObject.text[0].text).toBe('315\u00b0');
    });
  });
  describe('Figure Level', () => {
    test('Pad change shape', () => {
      figure.add({
        method: 'collections.polyline',
        name: 'a',
        options: {
          points,
          close: true,
          pad: {
            radius: 0.2,
            boundary: {
              translation: {
                left: -3, bottom: -2, right: 3, top: 2,
              },
            },
            isMovable: true,
          },
        },
      });
      figure.initialize();
      const a = figure.elements._a;
      expect(a.points[0]).toEqual(new Point(0, 0));
      expect(a.points[1]).toEqual(new Point(1, 0));
      expect(a.points[2]).toEqual(new Point(1, 1));
      const p0 = a._pad0;
      p0.setPosition(0, 0.5);
      expect(a.points[0]).toEqual(new Point(0, 0.5));
      expect(a.points[1]).toEqual(new Point(1, 0));
      expect(a.points[2]).toEqual(new Point(1, 1));
    });
    test('Pad Move freely along line shape', () => {
      figure.add({
        method: 'opolyline',
        name: 'a',
        options: {
          points: [
            [0, 0],
            [10, 0],
            [10, 10],
          ],
          close: true,
          pad: {
            radius: 0.2,
            boundary: {
              translation: {
                left: -3, bottom: -2, right: 3, top: 2,
              },
            },
            isMovable: true,
            // move: {
            //   bounds: {
            //     translation: { line: [[0, 0], [9, 9]] },
            //   },
            // },
          },
        },
      });
      figure.initialize();
      const a = figure.elements._a;
      // a.move.freely.bounceLoss = 0;
      // a.move.freely.deceleration = 1;
      expect(a.points[0]).toEqual(new Point(0, 0));
      expect(a.points[1]).toEqual(new Point(10, 0));
      expect(a.points[2]).toEqual(new Point(10, 10));
      const p0 = a._pad0;
      p0.move.bounds.updateTranslation({ line: [[0, 0], [9, 9]] });
      p0.setTouchable();
      p0.move.freely.bounceLoss = 0;
      p0.move.freely.deceleration = 1;
      figure.mock.timeStep(0);
      figure.mock.touchDown([0, 0]);
      figure.mock.timeStep(0.1);
      figure.mock.touchMove([0.5 * Math.sqrt(2), 0]);
      figure.mock.touchUp();

      // Moving at 5 units/s
      // Total time = 5s
      // Distance: s = v0t - 0.5*1*t^2 = 25 - 12.5 = 12.5;
      const x = 0.5 / Math.sqrt(2);
      expect(round(rectToPolar(p0.state.movement.velocity.t()).mag, 3)).toBe(5);
      expect(p0.getPosition().round(3).x).toEqual(round(x, 3));
      expect(round(p0.getRemainingMovingFreelyTime(), 2)).toBe(5);
      expect(p0.state.isMovingFreely).toBe(true);
      expect(a.points[0].round(3)).toEqual(new Point(x, x).round(3));
      expect(a.points[1]).toEqual(new Point(10, 0));
      expect(a.points[2]).toEqual(new Point(10, 10));

      // After 1s:
      //  v1 = v0 - at = 5 - 1 = 4
      //  s  = 5 - 0.5 = 4.5
      figure.mock.timeStep(1);
      expect(round(rectToPolar(p0.state.movement.velocity.t()).mag, 3)).toBe(4);
      expect(p0.getPosition().round(3).x).toEqual(round(x + 4.5 / Math.sqrt(2), 3));
      expect(round(p0.getRemainingMovingFreelyTime(), 2)).toBe(4);
      expect(p0.state.isMovingFreely).toBe(true);
      expect(a.points[0].round(3)).toEqual(new Point(
        x + 4.5 / Math.sqrt(2),
        x + 4.5 / Math.sqrt(2),
      ).round(3));
      expect(a.points[1]).toEqual(new Point(10, 0));
      expect(a.points[2]).toEqual(new Point(10, 10));

      // After 2s:
      //  v2 = v1 - at = 4 - 1 = 3
      //  s  = 4 - 0.5 = 3.5 (+4.5)
      figure.mock.timeStep(1);
      expect(round(rectToPolar(p0.state.movement.velocity.t()).mag, 3)).toBe(3);
      expect(p0.getPosition().round(3).x).toEqual(round(x + (4.5 + 3.5) / Math.sqrt(2), 3));
      expect(round(p0.getRemainingMovingFreelyTime(), 2)).toBe(3);
      expect(p0.state.isMovingFreely).toBe(true);
      expect(a.points[0].round(3)).toEqual(new Point(
        x + 8 / Math.sqrt(2),
        x + 8 / Math.sqrt(2),
      ).round(3));
      expect(a.points[1]).toEqual(new Point(10, 0));
      expect(a.points[2]).toEqual(new Point(10, 10));
    });
  });
});
