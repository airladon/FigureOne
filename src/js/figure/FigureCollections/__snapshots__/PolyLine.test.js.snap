// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Figure Objects PolyLine Misc options 1`] = `undefined`;

exports[`Figure Objects PolyLine Number of points in line close 1`] = `undefined`;
